src/librustc_typeck/check/pat.rs

```diff
@@ -371,17 +378,71 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            | {
+                if !actual_type.references_error() {
+                    let other_expr = other_expr.filter(|(_, ty)| !ty.references_error());
+                    let cause = self.pat_guard_cause(
+                        other_expr, cause_span, expected, discrim_span
+                    );
+
+                    let result = self.at(&cause, self.param_env).eq(expect_ty, actual_type);
+                    result.map_err(|terr| (cause, terr))
+                } else {
+                    Ok(InferOk {obligations: PredicateObligations::new(), value: ()})
+                }
+            };
+
+            // test if sides both sides of range have the compatible types
+            let lhs_result = at_cause_eq(expected, lhs_ty, begin.span, Some((end.span, rhs_ty)));
+            let rhs_result = at_cause_eq(expected, rhs_ty, end.span, Some((begin.span, lhs_ty)));
+
+            let joined_result = match (lhs_result, rhs_result) {
+                // full success, move forwards
+                (
+                    Ok(InferOk { obligations: lhs_obligation, value: () }),
+                    Ok(InferOk { obligations: rhs_obligation, value: () })
+                ) => {
+                    self.register_predicates(lhs_obligation);
+                    self.register_predicates(rhs_obligation);
+
+                    // Now that we know the types can be unified we find the unified type and use
+                    // it to type the entire expression.
+                    Ok(self.resolve_vars_if_possible(&lhs_ty))
+                },
+                // only lhs is wrong
+                (Err((cause, terr)), Ok(_)) =>
+                    Err(self.report_mismatched_types(&cause, expected, lhs_ty, terr)),
+                // only rhs is wrong
+                (Ok(_), Err((cause, terr))) =>
+                    Err(self.report_mismatched_types(&cause, expected, rhs_ty, terr)),
+                // both sides are wrong
+                (Err(_), Err((rhs_cause, terr))) => {
+                    if let Ok(_) = at_cause_eq(lhs_ty, rhs_ty, Span::default(), None) {
+                        let cause = self.pat_guard_cause(None, span, expected, discrim_span);
+                        Err(self.report_mismatched_types(&cause, expected, rhs_ty, terr))
+                    } else {
+                        Err(self.report_mismatched_types(&rhs_cause, expected, rhs_ty, terr))
+                    }
+                }
+            };
 
-        // Subtyping doesn't matter here, as the value is some kind of scalar.
-        self.demand_eqtype_pat(span, expected, lhs_ty, discrim_span);
-        self.demand_eqtype_pat(span, expected, rhs_ty, discrim_span);
-        Some(common_type)
+            match joined_result {
+                Ok(common_type) => Some(common_type),
+                Err(mut err) => {
+                    err.emit();
+                    None
+                }
+            }
+        }
```

Comment:
I'm not happy about the amount of special casing, rawness, and the fragility of this type checking logic.
Looking right now to see if there's a less complicated way to get near the same results.